website/content/specification/syntax/constraints.md

```diff
@@ -12,6 +12,49 @@ TODO: P3: Address issue https://github.com/usnistgov/metaschema/issues/325
 
 ## Common Constraint Data
 
+## Let expressions
+
+Using the `let` element, a variable can be defined, which can be used in a Metapath expression in subsequent constraints.
+
+A `let` statement has a REQUIRED `@var` attribute, which defines the variable name.
+
+A `let` statement has a REQUIRED `@expression` attributes, which defines an Metapath expression, whose result is used to define the variable's value in the evaluation context.
+
+During constraint evaluation, each `let` statement MUST be evaluated in encounter order. If a previous variable is bound with the same name in the evaluation context, the new value MUST bound in a sub-context to avoid side effects. This sub-context MUST be made available to any constraints following the `let` statement declaration, and to any constraints defined on child nodes of the current context.
+
+During evaluation, when a variable is bound for a `let` statement, the variables value MUST be set to the result of evaluating the `@expression` using the current node as the Metapath evaluation focus.
+
+For example:
+
+Given the following fragment of a Metaschema module.
+
+```xml
+<define-assembly name="sibling">
+  <define-flag name="name" required="yes"/>
+  <constraint>
+    <let var="parent" expression=".."/><!-- stores the parent of the current node -->
+    <let var="sibling-count" expression="count($parent/sibling)"/>
+    <expect target="." test="$sibling-count = 3"/>
+  </constraint>
+</define-assembly>
+```
+
+And the following document.
+
+```xml
+<parent name="p1">
+  <sibling name="a"/>
+  <sibling name="b"/>
+  <sibling name="c"/>
+</parent>
+<parent name="p2">
+  <sibling name="x"/>
+  <sibling name="Y"/>
+</parent1>
+```
+
+The expect constraint would pass for each `sibling` in the `parent` named "p1", and would fail for each `sibling` in the `parent` named "p2".
```

Comment:
Minor, non-blocking. I understand the intent but the "pass for each ..." and "fail for each ..." threw for a curve for a moment.

```suggestion
The expect constraint would pass for the `parent` named "p1" because the resulting `sibling-count` value is 3 as expected. It would fail for the `parent` named "p2" because the resulting `sibling-count` value is 2.
```

Reply:
The constraint is defined on the `sibling` node, so it would be evaluated on each `sibling` node, not the `parent` node. Given the test data, the constraint would be evaluated 5 times, once each for a-c, and x-y.